### PR TITLE
Invalid VAT id chars exception

### LINKED ISSUE #19: Check if non-latin chars are in the VAT id before validation
VAT id numbers like CY123123Μ come back from the VIES service as *not valid* because they don't even pass the regex mentioned in the WSDL.

> The vatNumber input parameter must follow the pattern [0-9A-Za-z\+\*\.]{2,12}

Because that's not the `Μ` you think it is, this one's a multibyte with hex 0xce9c. There's no reason to even try to VIES this, let's throw a nice exception instead.

Upstream bug https://github.com/mpociot/vat-calculator/issues/99

### PR DESCRIPTION
**Pull request by spaze**
Find invalid chars in the VAT number and throw an exception if there are some.

This should be merged when https://github.com/phpstan/phpstan-src/pull/324 is accepted and released (PHPStan dev dependency version in `composer.json` should be raised accordingly).

Fix #19

### CODE REVIEW

File: src/VatCalculator.php [hunk 1]
```diff
@@ -142,10 +144,17 @@ public function isValidVatNumber(string $vatNumber): bool
 	 * @return VatDetails
 	 * @throws VatCheckUnavailableException
 	 * @throws UnsupportedCountryException
+	 * @throws InvalidCharsInVatNumberException
 	 */
 	public function getVatDetails(string $vatNumber, ?string $requesterVatNumber = null): VatDetails
 	{
 		$vatNumber = str_replace([' ', "\xC2\xA0", "\xA0", '-', '.', ','], '', trim($vatNumber));
+
+		$invalidChars = preg_split('/[0-9A-Za-z+*.]+/', $vatNumber, -1, PREG_SPLIT_NO_EMPTY | PREG_SPLIT_OFFSET_CAPTURE);
```
Comment (spaze) [inline src/VatCalculator.php:153]: Should say where this regex comes from (it's from the WSDL).

File: .gitattributes [hunk 2]
```diff
@@ -1,7 +1,6 @@
 /.editorconfig export-ignore
 /.gitattributes export-ignore
 /.gitignore export-ignore
-/composer.lock export-ignore
 /phpstan.neon export-ignore
 /phpunit.xml.dist export-ignore
 /tests export-ignore
```
Comment (williamdes) [inline .gitattributes:6]: Just one note: please do export tests so people using your library can package and test it on debian like https://salsa.debian.org/php-team/pear/php-dasprid-enum/-/tree/main/test for example :) 

Just a random comment ;p

Comment (spaze) [reply .gitattributes:6]: Please file a new issue. I don't understand why tests should be exported when they're usually not ([e.g.](https://github.com/twigphp/Twig/blob/3.x/.gitattributes)) but feel free to change my mind. In a different issue, that is :-)

Comment (williamdes) [reply .gitattributes:6]: Hi,
I will not file an issue because I do not have the need for it, but to explain myself:
- When you are packaging a repository into Debian you will most probably use [a watch file](https://salsa.debian.org/phpmyadmin-team/mariadb-mysql-kbs/-/blob/main/debian/watch) that will configure scraping the tags or releases page.
- When a version is detected you can import it, there comes the vendor tarball.
- GitHub allows to download zips or tars from tags or releases and basically that does a `git archive tagNameHere`
- When you imported all the changes from the extracted zip like [here](https://salsa.debian.org/php-team/pear/php-dasprid-enum/-/commits/upstream) you will be able to configure things.
- After all the things seem okay you would want to provide what is called 'autopkgtests' and here is the point where you need the tests written on the lib. to be available to run: [example](https://salsa.debian.org/phpmyadmin-team/twig-i18n-extension/-/jobs/902218#L860).


Why https://github.com/twigphp/Twig/blob/3.x/.gitattributes is not a good example is because you can see by this [example](https://salsa.debian.org/php-team/pear/twig/-/commit/5cde1f50bb48a74e12b631631bf8ef40e2cbf2f3) that Twig packaging team merges the git history instead of importing tarballs.

For some reasons I hate this method because you can see [here](https://github.com/williamdes/mariadb-mysql-kbs/blob/main/.gitattributes) that I have quite a lot of files I do not want people to have on their computers.

Importing from git would bring me more problems that solutions, copyright issues to solve, etc...


TLDR; for me: keep your `git archive` as clean as possible, include source files and a way to test them.

Comment (spaze) [reply .gitattributes:6]: Aha, thanks. I don't really understand creating Debian packages, I think I have never created one (maybe except one local .deb file a long time ago in a galaxy far, far away). I don't see tests included in packages even though the original repository doesn't export-ignore them, e.g. [here](https://packages.debian.org/sid/all/php-dasprid-enum/filelist) so I still don't understand why they should be included and what difference it brings when using those exported tarballs instead of git repo.

In one of my `vendor` dirs, just 9 out of 64 packages include tests. I won't be adding it unless there's at least one person that needs it but I'd probably prefer them using the repo and unexport-ignoring tests instead of including tests in the archive. I might change my mind though.

Comment (williamdes) [reply .gitattributes:6]: > Aha, thanks. I don't really understand creating Debian packages

It is quite a "complex" process like learning how git works, personally I did not understand the how to manual until I did not need it "anymore" ^^

> I don't see tests included in packages even though the original repository doesn't export-ignore them, e.g. here 

Aha, this is because we choose witch files go where [here](https://salsa.debian.org/php-team/pear/php-dasprid-enum/-/blob/main/debian/install)

> so I still don't understand why they should be included and what difference it brings when using those exported tarballs instead of git repo.

One advantage is here https://tracker.debian.org/pkg/dasprid-enum "Required age reduced by 3 days because of autopkgtest" and also the ability to say a package works fine on an architecture, a package without tests is more difficult to say it works fine before really testing it. It is more a bundle of files that something alive if I can explain that like this. 

> In one of my vendor dirs, just 9 out of 64 packages include tests. I won't be adding it unless there's at least one person that needs it but I'd probably prefer them using the repo and unexport-ignoring tests instead of including tests in the archive. I might change my mind though.

Sure, I totally understand. Maybe someone will need them or will workaround this at packaging time ;)
One sure thing is to remove any non useful file from this vendor tarball, like phpstan configs, etc..

Comment (spaze) [reply .gitattributes:6]: PHPStan config is already export-ignored and I've just added .github and CS config. Please file a new issue if you think something else should (or shouldn't) be export-ignored as this is not a place for such things. Thanks.
